### PR TITLE
fix: modification of the setTranslation function for the case where the node is untranslated

### PR DESCRIPTION
**Pull request by smnperret-kdab**
refer to PR #87 for upcoming commits

### CODE REVIEW

File: src/core/qttsdocument.cpp [hunk 1]
```diff
@@ -297,10 +297,12 @@ QString QtTsMessage::translation() const
...
 
-void QtTsMessage::setTranslation(const QString &translate)
+void QtTsMessage::setTranslation(const QString &translation)
 {
-    LOG("QtTsMessage::setTranslation", translate);
-    m_message.child("translation").set_value(translate.toUtf8().constData());
+    LOG("QtTsMessage::setTranslation", translation);
+    m_message.remove_child("translation");
+    pugi::xml_node translationChild = m_message.append_child("translation"); // add translation
+    translationChild.append_child(pugi::node_pcdata).set_value(translation.toUtf8().constData());
```
Comment (narnaud) [inline src/core/qttsdocument.cpp:305]: I found the right way:
```
    m_message.child("translation").text().set(translation.toUtf8().constData());
```
No need to remove and re-add a node...

File: tests/tst_qttsdocument.cpp [hunk 2]
```diff
@@ -126,6 +126,18 @@ private slots:
         }
     }
 
+    void changeTranslation()
+    {
+        Core::QtTsDocument document;
+        document.load(Test::testDataPath() + QStringLiteral("/tst_qttsdocument/language_translation.ts"));
+        for (const auto &message : document.messages()) {
```
Comment (narnaud) [inline tests/tst_qttsdocument.cpp:133]: `std::as_const(document.messages())` to cancel clazy warning

File: tests/tst_qttsdocument.cpp [hunk 3]
```diff
@@ -126,6 +126,18 @@ private slots:
...
+            message->setTranslation("new translation");
+        }
+        for (const auto &message : document.messages()) {
```
Comment (narnaud) [inline tests/tst_qttsdocument.cpp:136]: Here too: `std::as_const(document.messages())` to cancel clazy warning

File: src/core/qttsdocument.cpp (hunk 1 above)
Comment (dfaure-kdab) [reply src/core/qttsdocument.cpp:305]: You'd have to remove the type attribute too (type="unfinished"). Not sure if that's simpler in the end. But text() certainly looks good.

File: tests/tst_qttsdocument.cpp (hunk 2 above)
Comment (dfaure-kdab) [reply tests/tst_qttsdocument.cpp:133]: Can't use as_const on an rvalue. This needs a const local variable.

File: src/core/qttsdocument.cpp [hunk 4]
```diff
@@ -297,10 +297,11 @@ QString QtTsMessage::translation() const
     return QString::fromUtf8(m_message.child("translation").text().as_string());
 }
 
-void QtTsMessage::setTranslation(const QString &translate)
+void QtTsMessage::setTranslation(const QString &translation)
 {
-    LOG("QtTsMessage::setTranslation", translate);
-    m_message.child("translation").set_value(translate.toUtf8().constData());
+    LOG("QtTsMessage::setTranslation", translation);
+    m_message.remove_child("translation");
```
Comment (narnaud) [inline src/core/qttsdocument.cpp:303]: I was unclear, please remove this line, only keep the line below.
Also add: `m_message.child("translation").remove_attribute("type");`

The tests will magically work if you do that. The reason they are failing is that you are removing the translation node, but you are not adding a new one: `m_message.child("translation")` returns an empty node and all the operations after that are NOOP.

**Comment by smnperret-kdab** [comment]
refer to PR #87 for upcoming commits

**Pull request closed without merging**